swift/extractor/SwiftDispatcher.h

```diff
@@ -0,0 +1,178 @@
+  // it actually gets emitted to handle recursive cases such as recursive calls, or recursive type
+  // declarations
+  template <typename E>
+  TrapLabel<ToTag<E>> assignNewLabel(E* e) {
+    assert(waitingForNewLabel == getCanonicalPtr(e) && "assignNewLabel called on wrong entity");
+    auto label = getLabel<ToTag<E>>();
+    trap.assignStar(label);
+    store.insert(e, label);
+    waitingForNewLabel = nullptr;
+    return label;
+  }
+
+  template <typename Tag>
+  TrapLabel<Tag> getLabel() {
+    return arena.allocateLabel<Tag>();
+  }
+
+  // This is a helper method to emit TRAP entries for AST nodes that we don't fully support yet.
+  template <typename Parent, typename Child>
+  void TBD(Child* entity, const std::string& suffix) {
+    using namespace std::string_literals;
+    auto label = assignNewLabel(entity);
+    auto kind = detail::getKindName<Parent>(static_cast<const Parent*>(entity)->getKind());
+    auto name = "TBD ("s + kind + suffix + ")";
+    if constexpr (std::is_same_v<Parent, swift::TypeBase>) {
+      trap.emit(UnknownTypesTrap{label, name});
+    } else {
+      trap.emit(UnknownAstNodesTrap{label, name});
+    }
+  }
+
+  template <typename Locatable>
+  void attachLocation(Locatable locatable, TrapLabel<LocatableTag> locatableLabel) {
+    attachLocation(&locatable, locatableLabel);
+  }
+
+  // Emits a Location TRAP entry and attaches it to an AST node
+  template <typename Locatable>
+  void attachLocation(Locatable* locatable, TrapLabel<LocatableTag> locatableLabel) {
+    auto start = locatable->getStartLoc();
+    auto end = locatable->getEndLoc();
+    if (!start.isValid() || !end.isValid()) {
+      // invalid locations seem to come from entities synthesized by the compiler
+      return;
+    }
+    std::string filepath = getFilepath(start);
+    auto fileLabel = arena.allocateLabel<FileTag>();
+    trap.assignKey(fileLabel, filepath);
+    /// TODO: do not emit duplicate trap entries for Files
+    trap.emit(FilesTrap{fileLabel, filepath});
+    auto [startLine, startColumn] = sourceManager.getLineAndColumnInBuffer(start);
+    auto [endLine, endColumn] = sourceManager.getLineAndColumnInBuffer(end);
+    auto locLabel = arena.allocateLabel<LocationTag>();
+    trap.assignKey(locLabel, '{', fileLabel, "}:", startLine, ':', startColumn, ':', endLine, ':',
+                   endColumn);
+    trap.emit(LocationsTrap{locLabel, fileLabel, startLine, startColumn, endLine, endColumn});
+    trap.emit(LocatablesTrap{locatableLabel, locLabel});
+  }
+
+  std::string getFilepath(swift::SourceLoc loc) {
+    /// TODO: this needs more testing
+    std::string displayName = sourceManager.getDisplayNameForLoc(loc).str();
+    llvm::SmallString<PATH_MAX> filePath(displayName);
+    if (std::error_code ec = llvm::sys::fs::make_absolute(filePath)) {
+      std::cerr << "Cannot make absolute path: '" << displayName << "': " << ec.message() << "\n";
+      return {};
+    }
+    llvm::sys::path::remove_dots(filePath);
```

Comment:
What's the goal of this function, and would it be better accomplished with something like `std::filesystem::path::lexically_normal` or `llvm::sys::fs::real_path`?

Reply:
good point, do we need resolving symlinks here or not? What is the behaviour for symlinked file locations in other extractors? [I've just asked that on slack](https://github.slack.com/archives/CP6990YTU/p1652429882632649)

`std::filesystem` is unfortunately not available yet as it's not supported by the current macOS deploy target we have

Reply:
I think `real_path` should do the job just fine.
The `TODO: this needs more testing` was supposed to mean "we need to think about more use cases to make the right decision on what this function should return" 😃

Reply:
So if you agree, I'd expand the TODO comment with these points (`CODEQL_PRESERVE_SYMLINKS` and checking whether case is handled correctly on case insensitive file systems), and maybe open an issue about it?

Reply:
It's worth filing an issue for that for future work!